File: backend/src/routes/approvals.ts
import type { FastifyInstance } from "fastify";

import type { ApprovalsRepository } from "../repositories/approvals-repository.js";
import { readNumberQuery, readStringArrayQuery, readStringQuery } from "./query-helpers.js";

export function registerApprovalRoutes(app: FastifyInstance, repository: ApprovalsRepository): void {
  app.get("/approvals", async (request) => {
    const query = request.query as Record<string, unknown>;
    return repository.list({
      fromMs: readNumberQuery(query.fromMs),
      toMs: readNumberQuery(query.toMs),
      sessionKey: readStringQuery(query.sessionKey),
      runId: readStringQuery(query.runId),
      riskLevel: readStringArrayQuery(query.riskLevel),
      limit: readNumberQuery(query.limit),
      cursor: readStringQuery(query.cursor),
      resolution: readStringQuery(query.resolution),
      toolName: readStringQuery(query.toolName),
      module: readStringQuery(query.module),
      scopeType: readStringQuery(query.scopeType),
      requesterOuId: readStringQuery(query.requesterOuId),
    });
  });

  app.get("/approvals/:approvalId", async (request, reply) => {
    const approvalId = readStringQuery((request.params as Record<string, unknown>).approvalId);
    const approval = approvalId ? repository.getById(approvalId) : null;
    if (!approval) {
      return reply.code(404).send({ ok: false, message: "Approval not found." });
    }
    return approval;
  });
}

File: backend/internal/routes/approvals.go
package routes

import (
	"github.com/gin-gonic/gin"

	"github.com/openclaw/lynx-guardian/backend/internal/httpserver"
	"github.com/openclaw/lynx-guardian/backend/internal/repo"
)

// RegisterApprovals mirrors registerApprovalRoutes.
func RegisterApprovals(router gin.IRoutes, repository *repo.ApprovalsRepository) {
	router.GET("/approvals", func(c *gin.Context) {
		values := c.Request.URL.Query()
		query := repo.ApprovalsListQuery{
			Q:             httpserver.ReadString(values, "q"),
			FromMs:        httpserver.ReadInt64(values, "fromMs"),
			ToMs:          httpserver.ReadInt64(values, "toMs"),
			SessionKey:    httpserver.ReadString(values, "sessionKey"),
			RunID:         httpserver.ReadString(values, "runId"),
			RiskLevel:     httpserver.ReadStringSlice(values, "riskLevel"),
			PageNum:       httpserver.ReadInt(values, "pageNum"),
			PageSize:      httpserver.ReadInt(values, "pageSize"),
			Limit:         httpserver.ReadInt(values, "limit"),
			Cursor:        httpserver.ReadString(values, "cursor"),
			Resolution:    httpserver.ReadString(values, "resolution"),
			ToolName:      httpserver.ReadString(values, "toolName"),
			Module:        httpserver.ReadString(values, "module"),
			ScopeType:     httpserver.ReadString(values, "scopeType"),
			RequesterOuID: httpserver.ReadString(values, "requesterOuId"),
		}

		page, err := repository.List(query)
		if err != nil {
			c.JSON(500, gin.H{"ok": false, "message": err.Error()})
			return
		}
		c.JSON(200, page)
	})

	router.GET("/approvals/:approvalId", func(c *gin.Context) {
		approvalID := c.Param("approvalId")
		if approvalID == "" {
			c.JSON(404, gin.H{"ok": false, "message": "Approval not found."})
			return
		}
		approval, err := repository.GetByID(approvalID)
		if err != nil {
			c.JSON(500, gin.H{"ok": false, "message": err.Error()})
			return
		}
		if approval == nil {
			c.JSON(404, gin.H{"ok": false, "message": "Approval not found."})
			return
		}
		c.JSON(200, approval)
	})
}
